addons/website/static/src/snippets/s_website_form/000.js

```diff
@@ -126,19 +124,36 @@ odoo.define('website.s_website_form', function (require) {
             // to know which user data should be used to prfill fields.
-            var $dataFor = $('[data-for=' + this.$target.attr('id') + ']');
-            if ($dataFor.length || !_.isEmpty(this.preFillValues)) {
-                var dataForValues = $dataFor.length ? JSON.parse($dataFor.data('values').replace('False', '""').replace('None', '""').replace(/'/g, '"')) : [];
-                var fields = _.pluck(this.$target.serializeArray(), 'name');
-                _.each(fields, function (field) {
-                    var $field = self.$target.find('input[name="' + field + '"], textarea[name="' + field + '"]');
-                    if (!$field.val() && _.has(dataForValues, field) && dataForValues[field]) {
-                        $field.val(dataForValues[field]);
-                    } else if (!$field.val() && $field.data('fill-with')) {
-                        $field.val(self.preFillValues[$field.data('fill-with')] || '');
+            const dataForEl = document.querySelector(`[data-for='${this.$target[0].id}']`);
+            if (dataForEl || Object.keys(this.preFillValues).length) {
+                const dataForValues = dataForEl ? JSON.parse(dataForEl.dataset.values
+                        .replace('False', '""').replace('None', '""').replace(/'/g, '"')) : {};
+                const fieldNames = this.$target.serializeArray().map(el => el.name);
+                this.initialValues = [];
+                // All types of inputs do not have a value property (eg:hidden),
+                // for these inputs any function that is supposed to put a value
+                // property actually puts a HTML value attribute. Because of
+                // this, we have to clean up these values at destroy or else the
+                // data loaded here could become default values. We could set
+                // the values to submit() for these fields but this could break
+                // customizations that use the current behavior as a feature.
+                for (const name of fieldNames) {
+                    const fieldEl = this.$target[0].querySelector(`[name="${name}"]`);
+                    let newValue;
+                    if (!fieldEl.value && dataForValues && dataForValues[name]) {
+                        newValue = dataForValues[name];
+                    } else if (!fieldEl.value && this.preFillValues[fieldEl.dataset.fillWith]) {
+                        newValue = this.preFillValues[fieldEl.dataset.fillWith];
                     }
-                    $field.data('website_form_original_default_value', $field.val());
-                });
+                    if (newValue) {
+                        this.initialValues.push({
+                            fieldEl,
+                            initialValue: fieldEl.getAttribute('value') || '',
+                            hasValueAttr: fieldEl.hasAttribute('value'),
+                        });
```

Comment:
Still not good enough: you now declare redundant information. Why do you need `initialValue` and `hasValueAttr`? Actually, if `initialValue` is falsy, then it's always better to remove the value attribute no? So it seems to be more code for a lesser behavior.

Also, use a `Map` as I suggested to map the field to its initial value. It makes way more sense than an array of objects.

Reply:
In fact I wanted to avoid that inputs like this `<input value="" />` or like this `<input value />` change but indeed it's probably not an issue if those inputs lost their value attribute

Reply:
It's more than not an issue, it's better :wink: